align/pnr/placer_pythonic.py

```diff
@@ -0,0 +1,116 @@
+import logging
+
+from align.schema.hacks import VerilogJsonTop
+from align.pnr.checker import check_placement
+from align.pnr.placer_pythonic_sp import place_using_sequence_pairs
+
+
+def compute_topoorder(modules, concrete_name, key='abstract_template_name'):
+    found_modules, found_leaves = set(), set()
+    order = list()
+
+    def aux(cn):
+        if cn in modules:
+            found_modules.add(cn)
+            for instance in modules[cn]['instances']:
+                ctn = instance[key]
+                if ctn not in found_modules and ctn not in found_leaves:
+                    aux(ctn)
+            order.append(cn)
+        else:
+            found_leaves.add(cn)
+            order.append(cn)
+    aux(concrete_name)
+
+    return order, found_modules, found_leaves
+
+
+def trim_placement_data(placement_data, top_level):
+
+    modules = {module['concrete_name']: module for module in placement_data['modules']}
+
+    top_concrete_names = [module['concrete_name'] for module in placement_data['modules'] if module['abstract_name'] == top_level]
+    all_modules_leaves = list()
+    for concrete_name in top_concrete_names:
+        _, found_modules, found_leaves = compute_topoorder(modules, concrete_name, key='concrete_template_name')
+        all_modules_leaves.extend(found_modules)
+        all_modules_leaves.extend(found_leaves)
+
+    all_modules_leaves = set(all_modules_leaves)
+
+    new_placement_data = {'leaves': list(), 'modules': list()}
+    new_placement_data['leaves'] = [leaf for leaf in placement_data['leaves'] if leaf['concrete_name'] in all_modules_leaves]
+    new_placement_data['modules'] = [leaf for leaf in placement_data['modules'] if leaf['concrete_name'] in all_modules_leaves]
+
+    for key in ['leaves', 'modules']:
+        for elem in new_placement_data[key]:
+            if 'pin_bbox' in elem:
+                del elem['pin_bbox']
+
+    return new_placement_data
+
+
+def propagate_down_global_signals(modules: dict, module_name: str, global_signals: list):
+    GS = 'global_signals'
+    modules[module_name][GS] = modules[module_name].get(GS, list()) + global_signals
+    for instance in modules[module_name]['instances']:
```

Comment:
It seems that we want to have a `global_signals` field in every submodule.
If there are going to be a lot of global signals, then we would want to use a set to hold them instead of a list because we are using `in` searches. We could use sets instead of list then while in this routine. (Convert from list to set at the beginning, then back from set to list can the end.)
Or ignore this, but probably mark it as a potential SLUG bug so we can fix it if it becomes a problem later.